tests/gtest/aviflosslesstest.cc

```diff
@@ -12,86 +14,127 @@ namespace {
+  const avifAppFileFormat file_format = avifReadImage(
+      file_path.c_str(),
+      /*requestedFormat=*/pixel_format,
+      /*requestedDepth=*/0,
+      /*chromaDownsampling=*/AVIF_CHROMA_DOWNSAMPLING_AUTOMATIC,
+      /*ignoreColorProfile=*/ignore_icc, /*ignoreExif=*/false,
+      /*ignoreXMP=*/false, /*allowChangingCicp=*/true,
+      /*ignoreGainMap=*/true, AVIF_DEFAULT_IMAGE_SIZE_LIMIT, image.get(),
+      /*outDepth=*/nullptr, /*sourceTiming=*/nullptr,
+      /*frameIter=*/nullptr);
+  if (matrix_coefficients == AVIF_MATRIX_COEFFICIENTS_YCGCO_RO) {
+    // AVIF_MATRIX_COEFFICIENTS_YCGCO_RO does not work because the input
+    // depth is not odd.
+    ASSERT_EQ(file_format, AVIF_APP_FILE_FORMAT_UNKNOWN);
+    return;
+  }
+  ASSERT_NE(file_format, AVIF_APP_FILE_FORMAT_UNKNOWN);
+
+  // Encode.
+  EncoderPtr encoder(avifEncoderCreate());
+  ASSERT_NE(encoder, nullptr);
+  encoder->speed = AVIF_SPEED_FASTEST;
+  encoder->quality = AVIF_QUALITY_LOSSLESS;
+  encoder->qualityAlpha = AVIF_QUALITY_LOSSLESS;
+  testutil::AvifRwData encoded;
+  avifResult result = avifEncoderWrite(encoder.get(), image.get(), &encoded);
+
+  if (matrix_coefficients == AVIF_MATRIX_COEFFICIENTS_IDENTITY &&
+      pixel_format != AVIF_PIXEL_FORMAT_NONE &&
+      pixel_format != AVIF_PIXEL_FORMAT_YUV444) {
+    // The AV1 spec does not allow identity with subsampling.
+    ASSERT_EQ(result, AVIF_RESULT_INVALID_ARGUMENT);
+    return;
+  }
+  ASSERT_EQ(result, AVIF_RESULT_OK) << avifResultToString(result);
+
+  // Decode to RAM.
+  ImagePtr decoded(avifImageCreateEmpty());
+  ASSERT_NE(decoded, nullptr);
+  DecoderPtr decoder(avifDecoderCreate());
+  ASSERT_NE(decoder, nullptr);
+  result = avifDecoderReadMemory(decoder.get(), decoded.get(), encoded.data,
+                                 encoded.size);
+  ASSERT_EQ(result, AVIF_RESULT_OK) << avifResultToString(result);
+
+  // What we read should be what we encoded
+  ASSERT_TRUE(testutil::AreImagesEqual(*image, *decoded));
+
+  // Convert to a temporary PNG and read back, to have default matrix
+  // coefficients.
+  const std::string tmp_path = testing::TempDir() + "decoded_default.png";
+  ASSERT_TRUE(testutil::WriteImage(decoded.get(), tmp_path.c_str()));
+  const ImagePtr decoded_lossless =
+      ReadImageLossless(tmp_path, pixel_format, ignore_icc);
+
+  // Verify that the ground truth and decoded images are the same.
+  const ImagePtr ground_truth_lossless =
+      ReadImageLossless(file_path, pixel_format, ignore_icc);
+  const bool are_images_equal =
+      testutil::AreImagesEqual(*ground_truth_lossless, *decoded_lossless);
+
+  if (matrix_coefficients == AVIF_MATRIX_COEFFICIENTS_YCGCO) {
+    // AVIF_MATRIX_COEFFICIENTS_YCGCO is not a lossless transform.
+    ASSERT_FALSE(are_images_equal);
+  } else if (pixel_format == AVIF_PIXEL_FORMAT_YUV400) {
+    // For gray images, information is lost.
+    ASSERT_FALSE(are_images_equal);
+  } else {
+    ASSERT_TRUE(are_images_equal);
```

Comment:
This looks unrelated to the test above because `ASSERT_TRUE(testutil::AreImagesEqual(*image, *decoded));`. Split into two units?